types/src/validator_verifier.rs

```diff
@@ -159,54 +166,99 @@ impl ValidatorVerifier {
-    /// attached signatures is invalid or it does not correspond to a known author. The latter is to
-    /// prevent malicious users from adding arbitrary content to the signature payload that would go
-    /// unnoticed.
-    pub fn verify_aggregated_struct_signature<T: CryptoHash + Serialize>(
+    pub fn generate_aggregated_signature(
         &self,
-        message: &T,
-        aggregated_signature: &BTreeMap<AccountAddress, bls12381::Signature>,
-    ) -> std::result::Result<(), VerifyError> {
-        self.check_num_of_signatures(aggregated_signature)?;
-        self.check_voting_power(aggregated_signature.keys())?;
-        for (author, signature) in aggregated_signature {
-            self.verify(*author, message, &signature.clone())?;
+        partial_signatures: &PartialSignatures,
+    ) -> Result<(AggregatedSignature, PublicKey), VerifyError> {
+        let validator_bitmask = self
+            .address_to_validator_info
+            .iter()
+            .map(|(address, _)| partial_signatures.signatures().contains_key(address))
+            .collect::<Vec<_>>();
+        // Perform an optimistic aggregation of the signatures.
+        let aggregated_sig = bls12381::Signature::aggregate(
+            partial_signatures.signatures().values().cloned().collect(),
+        )
+        .map_err(|_| VerifyError::FailedToAggregateKey)?;
+
+        // Verify the optimistically aggregated signature.
+        let mut pub_keys_to_agg = vec![];
+        for address in partial_signatures.signatures().keys() {
+            pub_keys_to_agg.push(
+                self.address_to_validator_info
+                    .get(address)
+                    .ok_or(VerifyError::UnknownAuthor)?
+                    .public_key(),
+            );
         }
-        Ok(())
+
+        let aggregated_key =
+            PublicKey::aggregate(pub_keys_to_agg).map_err(|_| VerifyError::FailedToAggregateKey)?;
+        Ok((
+            AggregatedSignature::new(validator_bitmask, Some(aggregated_sig)),
+            aggregated_key,
+        ))
+    }
+
+    pub fn verify_and_generate_aggregated_signature<T: CryptoHash + Serialize>(
+        &self,
+        partial_signatures: &PartialSignatures,
+        message: &T,
+    ) -> Result<AggregatedSignature, VerifyError> {
+        let (aggregated_sig, aggregated_key) =
+            self.generate_aggregated_signature(partial_signatures)?;
+        aggregated_sig
+            .aggregated_signature()
+            .as_ref()
+            .ok_or(VerifyError::FailedToAggregateKey)?
+            .verify(message, &aggregated_key)
+            .expect("Failed to verify the aggregated bls signature");
+        Ok(aggregated_sig)
     }
 
-    /// This function will try batch signature verification and falls back to normal
-    /// iterated verification if batching fails.
-    pub fn batch_verify_aggregated_signatures<T: CryptoHash + Serialize>(
+    pub fn verify_aggregated_signatures<T: CryptoHash + Serialize>(
```

Comment:
should be called verify_multi_signatures, we'll have another verify_aggregated_signatures with &[&T]